text/0000-crate-changelog-field.md

```diff
@@ -0,0 +1,228 @@
+
+- This proposal does not check if the `changelog` field points to a valid
+  external changelog, requiring crate authors to do that check themselves.
+- This proposal does not require the changelog to be present in the crate,
+  making offline usage of the `chagelog` field impractical.
+- This proposal does not constrain the format of the changelog, making it hard
+  for external tools to consume the contents of the changelog.
+
+# Rationale and Alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Rationale
+[rationale]: #rationale
+
+Most of the time, the producer of changelogs are crate authors
+and the consumer of changelogs are other crate authors.
+For human consumption, it is ideal to have the changelog in a rendered format.
+But handling rendering of various formats in crates.io is a heavy task.
+However, many existing crates already have changelogs available in their
+repository and the websites backing their repositories can already render the
+changelogs. So, in order to avoid unnecessary efforts, this RFC proposes to
+just use a valid URL for the `changelog` field.
+
+## Alternative: Local File
+[alternative-local-file]: #alternative-local-file
+
+The `changelog` field can be a path to a file within the crate,
+which opens the possibility for crates.io to render it directly.
+However, people have different opinions when it comes to what format the
+changelog should use. While markdown is currently the most popular choice,
+plain text and reStructuredText are also in use.
+Instead of handling the rendering of various different formats of changelog in
+crates.io, it is much easier to just have an external link, where an external
+website handles the rendering of the changelog so that crate authors can use
+any format they see fit for their changelogs.
+
+And changelogs grow over time, which can take significant amount of disk spaces
+for crates.io, considering that usually every single release of a crate
+will contain all changelogs up to that release.
+This particular drawback can be solved by introducing new linting rules.
+
+From another perspective, using an external link can guarantee that we can
+always link to the latest changelog, which is supposed to contain all
+changelog entries from initial versions to unreleased commits.
+If we adopt the local file approach, when we view an outdated version of a
+crate on docs.rs, we only get a link to the outdated changelog, which is not
+helpful for viewing the breaking changes introduced in new versions, requring
+another click on the doc.rs page to switch to the latest version.
+
+## Alternative: Do Nothing
+[alternative-do-nothing]: #alternative-do-nothing
+
+It has been several years after the initial changelog RFC.
+We can continue to live without this enhancement.
+
+It is possible to create an external registry that collects the links of
+changelogs. However, such effort is neither scalable nor ideal.
+
+# Prior art
+[prior-art]: #prior-art
```

Comment:
Are there other ecosystems that provide this? How do they handle it?  Are they used effectively?

Reply:
I haven't consider that point. Thanks for mentioning it! I will check other ecosystems.

Reply:
I don't know if that counts as ecosystems, but Debian packages have changelogs (sometimes left deliberately empty) and they can be referred to between versions using `apt` tools.

However, knowing burden of maintaining them (each new release require commit with new changelog effectively for it to be true), would be nice if `cargo publish` have option to generate own one based on git history (this could be extension).

Reply:
I assume Debian's isn't the complete changelog but a fragment exclusive for that specific release?  Handling situations like that is one of the limitations the previous RFC ran into.

Reply:
> would be nice if cargo publish have option to generate own one based on git history (this could be extension).

FWIW, I believe that VCS histories make terrible changelogs (in general). Of note is the inability to remove items or edit them once they have been merged (without rewriting public history). Forges (except Reviewboard and Gerrit) making commit messages not part of review also make them easy to not put into the review workflow like normal content does.

Reply:
@epage https://manpages.debian.org/testing/dpkg-dev/deb-changelog.5.en.html#debian/changelog

The file can have multiple entries, many packages have very long history.

@mathstuf point taken, however, I proposed option; then GitHub recently let merge commit to be titled from PR title, these are reviewed, full PR description can be included, but usually first line of commit lands in the change log.

Reply:
PyPI supports showing a user defined table of URLs, which is more extensible:

https://packaging.python.org/en/latest/specifications/well-known-project-urls/#well-known-labels

Reply:
Since this RFC doesn't require any specific format. It gives freedom to use any changelog-generating tooling people want. Some people like [git-cliff](https://git-cliff.org/), some may just write a bit manually, some can just link to their GitHub releases page.

Reply:
> In the R ecosystem we must use a file called `NEWS` (or `NEWS.md`). Both have style conventions, but they’re not enforced.
> 
> It has integrations into the R standard library (functions for interactively pulling up the changelog) and is displayed frequently in repository web frontends.
> 
> I’d estimate that a majority of popular packages use it.

@dgkf Thank you. I have added this to prior arts section.